### PR TITLE
FIX: move the font lock higher up the call and class tree

### LINKED ISSUE #26289: [Bug]: mathtext caching issue in multi-threaded environment with tight_layout=True
### Bug summary

Encountered the issue when plotting images with mathematical notation and `tight_layout=True` in Django - it occasionally crashes with the traceback suggesting some text metric cache corruption.  

The minimal working example below does not crash anymore if I remove the mathematical notation (`$\pm$` in the example), or use `tight_layout=False`

### Code for reproduction

```python
from matplotlib.figure import Figure

import io
import threading

def test_crash2():
    for i in range(100):

        fig = Figure(tight_layout=True)
        # fig = Figure()

        ax = fig.add_subplot(1, 1, 1)

        ax.text(0, 0, 'test $\pm$ 1.2') # This crashes
        # ax.text(0, 0, 'test 1.2') # This does not

        buf = io.BytesIO()
        fig.savefig(buf, format='png')

if __name__ == '__main__':

    for i in range(10):
        x = threading.Thread(target=test_crash2)
        x.start()
```

### Actual outcome

It produces the following traceback, essentially identical to what I saw in Django:
```
Traceback (most recent call last):
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/karpov/sciwork/stdweb/stdweb/test_crash.py", line 21, in test_crash2
    fig.savefig(buf, format='png')
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/figure.py", line 3343, in savefig
    self.canvas.print_figure(fname, **kwargs)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/backend_bases.py", line 2342, in print_figure
    self.figure.draw(renderer)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/artist.py", line 95, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/figure.py", line 3134, in draw
    self.get_layout_engine().execute(self)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/layout_engine.py", line 178, in execute
    kwargs = get_tight_layout_figure(
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/_tight_layout.py", line 266, in get_tight_layout_figure
    kwargs = _auto_adjust_subplotpars(fig, renderer,
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/_tight_layout.py", line 82, in _auto_adjust_subplotpars
    bb += [martist._get_tightbbox_for_layout_only(ax, renderer)]
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/artist.py", line 1417, in _get_tightbbox_for_layout_only
    return obj.get_tightbbox(*args, **kwargs)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 4408, in get_tightbbox
    bbox = a.get_tightbbox(renderer)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/artist.py", line 367, in get_tightbbox
    bbox = self.get_window_extent(renderer)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/text.py", line 959, in get_window_extent
    bbox, info, descent = self._get_layout(self._renderer)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/text.py", line 386, in _get_layout
    w, h, d = _get_text_metrics_with_cache(
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/text.py", line 97, in _get_text_metrics_with_cache
    return _get_text_metrics_with_cache_impl(
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/text.py", line 105, in _get_text_metrics_with_cache_impl
    return renderer_ref().get_text_width_height_descent(text, fontprop, ismath)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/backends/backend_agg.py", line 230, in get_text_width_height_descent
    self.mathtext_parser.parse(s, self.dpi, prop)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/mathtext.py", line 226, in parse
    return self._parse_cached(s, dpi, prop)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/mathtext.py", line 247, in _parse_cached
    box = self._parser.parse(s, fontset, fontsize, dpi)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/_mathtext.py", line 1992, in parse
    result = self._expression.parseString(s)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/pyparsing/core.py", line 1131, in parse_string
    loc, tokens = self._parse(instring, 0)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/pyparsing/core.py", line 921, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/pyparsing/core.py", line 817, in _parseNoCache
    loc, tokens = self.parseImpl(instring, pre_loc, doActions)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/pyparsing/core.py", line 3864, in parseImpl
    loc, resultlist = self.exprs[0]._parse(
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/pyparsing/core.py", line 921, in _parseCache
    value = self._parseNoCache(instring, loc, doActions, callPreParse)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/pyparsing/core.py", line 856, in _parseNoCache
    tokens = fn(instring, tokens_start, ret_tokens)
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/pyparsing/core.py", line 291, in wrapper
    ret = func(*args[limit:])
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/_mathtext.py", line 2028, in non_math
    symbols = [Char(c, self.get_state()) for c in s]
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/_mathtext.py", line 2028, in <listcomp>
    symbols = [Char(c, self.get_state()) for c in s]
  File "/Users/karpov/miniconda3/envs/stdweb/lib/python3.10/site-packages/matplotlib/_mathtext.py", line 2005, in get_state
    return self._state_stack[-1]
TypeError: 'NoneType' object is not subscriptable
```

### Expected outcome

No failures

### Additional information

_No response_

### Operating system

Tested on OS/X and on latest Ubuntu

### Matplotlib Version

3.7.1 and 3.6.3

### Matplotlib Backend

_No response_

### Python version

3.10.11 and 3.11.2

### Jupyter version

_No response_

### Installation

None

### PR DESCRIPTION
**Pull request by tacaswell**
## PR summary

We have for a long time (~2012) had an RLock on `RenderAgg` that was used in `FigureCanvasAgg.draw` to protect the shared cache of ft2font objects (the underlying c++ is very stateful and not thread safe). This lock was also implicitly protecting the mathtext cache when using the Agg backend.

However, given the recent improvements to the layout code there are now ways to call call `Figure.draw(renderer)` without acquiring this lock which leads to exceptions when using mathtext and a layout manager to save independent figures on different threads.

This bug also exists for the other backendends which use both the mathtext parser and the ft2font cache in the case of rendering texts as paths (in the vector backends).

The fix is to:

 - pull the lock up to `RendererBase` so all renderer instances share a single lock
 - acquire the lock in `Figure.draw` which is always the top entry point to rendering a Figure

Closes #26289

Not sure how to test this.  I guess we could add a test like:

```python
from matplotlib.figure import Figure

import io
import threading


def test_crash2():
    for i in range(100):
        fig = Figure(tight_layout=True)
        # fig = Figure()

        ax = fig.add_subplot(1, 1, 1)

        ax.text(0, 0, "test $\pm$ 1.2")  # This crashes
        # ax.text(0, 0, 'test 1.2') # This does not

        buf = io.BytesIO()
        fig.savefig(buf, format="svg")


if True:
    threads = [threading.Thread(target=test_crash2) for _ in range(10)]
    [t.start() for t in threads]
    [t.join() for t in threads]

```

but it seems a bit of a dice roll if you put in enough iterations to feel safe you never lost the race but few enough it does not take too long.

If this is an acceptable change, I think we should also add a section to the docs clarifying what level of threading we think we support and what we do not (one thread per figure should work in my opinion).

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

File: lib/matplotlib/figure.py [hunk 1]
```diff
@@ -3125,32 +3125,32 @@ def clear(self, keep_observers=False):
     def draw(self, renderer):
         # docstring inherited
 
-        # draw the figure bounding box, perhaps none for white figure
-        if not self.get_visible():
-            return
+        with getattr(renderer, "lock", nullcontext()):
```
Comment (anntzer) [inline lib/matplotlib/figure.py:3128]: As argued elsewhere, I think we should either just document that third-party renderers must inherit RendererBase (and thus skip the getattr here), or at least add a check to warn that renderers with no lock attribute are deprecated.

Comment (tacaswell) [reply lib/matplotlib/figure.py:3128]: That is the point.

I'm thinking about the best place to put that check.

Comment (tacaswell) [reply lib/matplotlib/figure.py:3128]: I mean, the other option is we stick the lock on `Figure` on on some module.

The other analogy I know of a global-ish lock like this is the global lock in h5py to protect calling into libhdf5: https://github.com/h5py/h5py/blob/6b5af4c6495bf865fee5f036122187c21fcb17d4/h5py/_objects.pyx#L40-L46 .

Expecting the instance to carry the lock like this is "nice" in that things seem well encapsulated, but is leaves us open to some backend opting out (or using a different lock) and bringing back a super subtle version of this bug ("this only happens when I save a mix of svg and png in a multi-threaded environment....").

I've talked my self into making this a private module level attribute on `Figure`.

We can make it public later if we need to.

**Comment by tacaswell** [comment]
I also added a test for one of the changed lines that was not previously covered.

There are obviously other ways to get at the caches (I suspect if you ask a `Text` object how big it is outside of `draw` / `draw_without_rendering` you can still get your self in trouble but that was already the case and this at least gets the "standard" paths covered.

File: lib/matplotlib/figure.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/figure.py]: Sure, on the figure seems fine, too.

**Comment by tacaswell** [comment]
All of the failures are codecov uploads failing.

File: lib/matplotlib/figure.py [hunk 2]
```diff
@@ -3125,32 +3138,32 @@ def clear(self, keep_observers=False):
     def draw(self, renderer):
         # docstring inherited
 
-        # draw the figure bounding box, perhaps none for white figure
-        if not self.get_visible():
-            return
+        with self._render_lock:
+            if not self.get_visible():
+                return
```
Comment (greglucas) [inline lib/matplotlib/figure.py:3143]: Leave this outside the lock to save acquiring the lock and releasing right away?

File: lib/matplotlib/figure.py [hunk 3]
```diff
@@ -2365,6 +2366,18 @@ class Figure(FigureBase):
...
 
+    # we want to cache the fonts and mathtext at a global level so that when
+    # multiple figures are created we can reuse them.  This helps with a bug on
+    # windows where the creation of too many figures leads to too many open
+    # file handles and improves the performance of parsing mathtext.  However,
+    # these global caches are not thread safe.  The solution here is to let the
+    # Figure acquire a shared at the start of the draw, and release it when it
+    # is done.  This allows multiple renderers to share the cached fonts and
+    # parsed text, but only one figure can draw at time and so the font cache
+    # and mathtext cache are used by only one renderer at a time.
```
Comment (greglucas) [inline lib/matplotlib/figure.py:2377]:
```suggestion
    # Figure acquire a shared lock at the start of the draw, and release it when it
    # is done.  This allows multiple renderers to share the cached fonts and
    # parsed text, but only one figure can draw at a time and so the font cache
    # and mathtext cache are used by only one renderer at a time.
```

**Review by greglucas — APPROVED** [review]

**Review by ksunden — APPROVED** [review]

**Pull request merged**